### DISCUSSION TITLE
Using feature extraction layers

### DISCUSSION THREAD

**Discussion by desh2608** [lhotse-speech/lhotse#455] [Q&A] [answered]
I am trying to use `lhotse.features.kaldi.layers.Wav2Spec` as layer in a network as follows:

```python
feat_extractor = lhotse.features.kaldi.layers.Wav2Spec()

x = feat_extractor.forward(x)
```

What should be the shape of the input `x` to  the `forward()` method?

**Comment by csukuangfj** [comment, marked as answer]
As far as I can tell, the input shape of x should be
```
(batch_size, num_samples)
```

However, there is incorrect documentation in the code.

---

You can find the correct shape for `x` by reading the following code:

https://github.com/lhotse-speech/lhotse/blob/9b6e3e454667c067a2de70721cb771b02d1ae3b2/lhotse/features/kaldi/layers.py#L297

https://github.com/lhotse-speech/lhotse/blob/9b6e3e454667c067a2de70721cb771b02d1ae3b2/lhotse/features/kaldi/layers.py#L135-L151

You can see from line 143, `x` is at least a 2-D tensor.

https://github.com/lhotse-speech/lhotse/blob/9b6e3e454667c067a2de70721cb771b02d1ae3b2/lhotse/features/kaldi/layers.py#L580-L595

Line 594 and 595 give you the shape you wanted.

BTW, the documentation at line 581 is incorrect. `x` is not a 1-D tensor, but a 2-D tensor.
https://github.com/lhotse-speech/lhotse/blob/9b6e3e454667c067a2de70721cb771b02d1ae3b2/lhotse/features/kaldi/layers.py#L581

https://github.com/lhotse-speech/lhotse/blob/master/lhotse/features/kaldi/layers.py
needs to improve the documentation, I believe.

**Reply by desh2608** [reply]
Thanks @csukuangfj. So I guess multi-channel tensors are not handled at the moment?

**Reply by csukuangfj** [reply]
How about viewing batch size dim as channel dim so that you can extract features for each channel independently?

**Reply by desh2608** [reply]
Thanks, that makes sense. A desirable addition for the future may be spatial features (e.g. IPD).

**Reply by pzelasko** [reply]
Thanks for reporting this, I'll try to find some time to fix the docs. 

Open to PRs with new features :)

**Reply by pzelasko** [reply]
The documentation issues should be solved in #467
